File: backend/internal/invoices/money.go
package invoices

import (
	"strconv"
	"strings"

	"github.com/Alisher24/CarSharing/backend/internal/billing"
)

// How an amount is written for a person: the som a sum of tyiyn makes, the two minor digits below it,
// and the unit the interface names. The rule is the one the browser writes a price by
// (`frontend/src/features/fleet/money.ts`), stated here because a process cannot import it.
const (
	// tyiynInSom is how many tyiyn make one som.
	tyiynInSom = 100

	// moneySuffix is what follows an amount, in the unit the interface writes it in: an amount is
	// written as `12,34 сома` and never as a bare number.
	moneySuffix = " сома"

	// minorDigitsCount is how many digits below the som are always printed, so an amount is written
	// as `12,00` rather than as `12`.
	minorDigitsCount = 2

	// digitGroup is how many digits stand between two group separators.
	digitGroup = 3

	// groupSeparator is what separates the groups. The browser locale inserts a non-breaking space
	// where this letter writes an ordinary one, which is the one difference between the two: the
	// letter is plain text a person reads in a terminal or a mailbox, and no reader of it separates
	// the groups again.
	groupSeparator = ' '
)

// somText renders a sum of whole tyiyn the way the interface writes money: the som it makes, a comma
// and the two minor digits, followed by the unit. The digits are computed on whole numbers, so a sum
// beyond the exact range of a floating-point number arrives digit for digit.
func somText(amount billing.AmountTyiyn) string {
	whole := int64(amount) / tyiynInSom
	return groupedDigits(whole) + "," + minorDigits(int64(amount)%tyiynInSom) + moneySuffix
}

// groupedDigits writes a whole number of som, grouping its digits from the right.
func groupedDigits(value int64) string {
	digits := strconv.FormatInt(value, 10)
	var grouped strings.Builder
	for position, digit := range digits {
		if position > 0 && (len(digits)-position)%digitGroup == 0 {
			grouped.WriteRune(groupSeparator)
		}
		grouped.WriteRune(digit)
	}
	return grouped.String()
}

// minorDigits writes the digits below one som, which are printed even when they are zero.
func minorDigits(value int64) string {
	minor := strconv.FormatInt(value, 10)
	for len(minor) < minorDigitsCount {
		minor = "0" + minor
	}
	return minor
}

File: scripts/acceptance/money.mjs
// How an amount of the contract is written for a person: whole tyiyn, both minor digits, and the unit
// the interface writes. It is the rule `frontend/src/shared/money.ts` applies, restated here because a
// browser check compares a letter and a screen against an invoice the service published, and a check
// that read the expected text out of the thing it checks would prove nothing.
//
// The two parts of the rule a check cannot derive from the amount it is checking — the locale, and the
// unit an amount is written with — are read from that declaration by `scripts/money-declaration.test.mjs`
// rather than assumed: a change to either is a change here, not a check that compares against a unit
// nobody writes.
const INTERFACE_LOCALE = 'ru-RU';

/** How many tyiyn make one som. */
const TYIYN_IN_SOM = 100n;

/** Som are written with both minor digits, as `12,34` rather than `12,3`. */
const MINOR_DIGITS = 2;

/** The unit the interface writes an amount with, which is the som and not `KGS`. */
const SOM_UNIT = 'сома';

const somFormat = new Intl.NumberFormat(INTERFACE_LOCALE);
const DECIMAL_MARK = somFormat.formatToParts(1.1).find((part) => part.type === 'decimal')?.value ?? ',';

/** One amount of whole tyiyn written the way a person reads it. */
export function somText(tyiyn) {
  const amount = BigInt(tyiyn);
  const minor = (amount % TYIYN_IN_SOM).toString().padStart(MINOR_DIGITS, '0');
  return `${somFormat.format(amount / TYIYN_IN_SOM)}${DECIMAL_MARK}${minor} ${SOM_UNIT}`;
}
